src/sentry/models/notificationaction.py

```diff
@@ -0,0 +1,175 @@
+
+    @classmethod
+    def get_value(cls, name: str) -> int:
+        invert_choices = {v: k for k, v in dict(cls.as_choices()).items()}
+        return invert_choices[name]
+
+
+class ActionServiceType(FlexibleIntEnum):
+    """
+    The available services to fire action notifications
+    """
+
+    EMAIL = 0
+    PAGERDUTY = 1
+    SLACK = 2
+    MSTEAMS = 3
+    SENTRY_APP = 4
+    SENTRY = 5  # Use notification platform (src/sentry/notifications)
+
+    @classmethod
+    def as_choices(cls) -> Iterable[Tuple[int, str]]:
+        return (
+            (cls.EMAIL, "email"),
+            (cls.PAGERDUTY, "pagerduty"),
+            (cls.SLACK, "slack"),
+            (cls.MSTEAMS, "msteams"),
+            (cls.SENTRY_APP, "sentry_app"),
+            (cls.SENTRY, "sentry"),
+        )
+
+
+class ActionTargetType(FlexibleIntEnum):
+    """
+    Explains the contents of target_identifier
+    """
+
+    # The target_identifier is a direct reference used by the service (e.g. email address, slack channel id)
+    SPECIFIC = 0
+    # The target_identifier is an id from the User model in Sentry
+    USER = 1
+    # The target_identifier is an id from the Team model in Sentry
+    TEAM = 2
+    # The target_identifier is an id from the SentryApp model in Sentry
+    SENTRY_APP = 3
+
+    @classmethod
+    def as_choices(cls) -> Iterable[Tuple[int, str]]:
+        return (
+            (cls.SPECIFIC, "specific"),
+            (cls.USER, "user"),
+            (cls.TEAM, "team"),
+            (cls.SENTRY_APP, "sentry_app"),
+        )
+
+
+class AbstractNotificationAction(Model):
+    """
+    Abstract model meant to retroactively create a contract for notification actions
+    (e.g. metric alerts, spike protection, etc.)
+    """
+
+    integration = FlexibleForeignKey("sentry.Integration", null=True)
+    sentry_app = FlexibleForeignKey("sentry.SentryApp", null=True)
+
+    # The type of service which will receive the action notification (e.g. slack, pagerduty, etc.)
+    type = BoundedIntegerField(choices=ActionServiceType.as_choices())
+    # The type of target which the service uses for routing (e.g. user, team)
+    target_type = BoundedIntegerField(choices=ActionTargetType.as_choices())
+    # Identifier of the target for the given service (e.g. slack channel id, pagerdutyservice id)
+    target_identifier = models.TextField(null=True)
+    # User-friendly name of the target (e.g. #slack-channel, pagerduty-service-name)
+    target_display = models.TextField(null=True)
+
+    @property
+    def service_type(self) -> ActionServiceType:
+        """
+        Used for disambiguity of self.type
+        """
+        return self.type
+
+    class Meta:
+        abstract = True
+
+
+class ActionTriggerType(FlexibleIntEnum):
+    """
+    The possible sources of action notifications
+    """
+
+    SPIKE_PROTECTION = 0  # Handlers registered in getsentry
+    SPEND_ALLOCATIONS = 1  # Handlers registered in getsentry
```

Comment:
Neither of these are in sentry, should we have a way for getsentry to register new ActionTriggerTypes during application startup? It always feels awkward to have features spread across getsentry & sentry like this.

Perhaps we could have an API like `ActionTriggerType.add_choice(0, 'SPIKE_PROTECTION', 'spike_protection')` that adds a new choice and enum property? Defining type choices at runtime means we would lose some benefits of typing as the valid attributes are no longer available statically.

Reply:
Addressed with the `TriggerGenerator()`. Tested it locally and it works for extending the Django validation :).


```python
NotificationAction.register_trigger_type(100, 'spike-protection')
NotificationAction.register_trigger_type(101, 'spend-allocations')
```

I had trouble trying to extend the enum directly though, so there is going to be a separate enum in getsentry
```python
class GetsentryActionTriggerType(IntEnum):

    SPIKE_PROTECTION = 100
    SPEND_ALLOCATIONS = 101

    @classmethod
    def as_choices(cls) -> Iterable[Tuple[int, str]]:
        return (
            (cls.SPIKE_PROTECTION, "spike-protection"),
            (cls.SPEND_ALLOCATIONS, "spend-allocations"),
        )
```